File: wasm/src/ast/_expr.ts
// _expr.ts — Expression hierarchy. Mirrors `ast::Expr` in `crates/ast/src/ast.rs`.
//
// Rust uses `#[serde(tag = "type")]` so each variant appears in JSON as
// `{"type": "<Variant>", ...}`. `exprFromWire` dispatches on that tag.

import { nodeFromWire } from "./_base";
import * as _dto from "./_dto";
import * as _types from "./_types";

function targetFromWire(w: Record<string, unknown>): unknown {
  return _dto.targetFromWire(Object.assign({}, w, { type: undefined }));
}

function identifierFromWire(w: Record<string, unknown>): unknown {
  return {
    names: ((w.names as unknown[]) || []).map((n) => nodeFromWire(n as never, (x: string) => x)),
    pkgpath: w.pkgpath || "",
    ctx: w.ctx,
  };
}

function unaryFromWire(w: Record<string, unknown>): unknown {
  return {
    op: w.op,
    operand: nodeFromWire(w.operand as never, exprFromWire as never),
  };
}

function binaryFromWire(w: Record<string, unknown>): unknown {
  return {
    left: nodeFromWire(w.left as never, exprFromWire as never),
    op: w.op,
    right: nodeFromWire(w.right as never, exprFromWire as never),
  };
}

function ifFromWire(w: Record<string, unknown>): unknown {
  return {
    body: nodeFromWire(w.body as never, exprFromWire as never),
    cond: nodeFromWire(w.cond as never, exprFromWire as never),
    orelse: nodeFromWire(w.orelse as never, exprFromWire as never),
  };
}

function selectorFromWire(w: Record<string, unknown>): unknown {
  return {
    value: nodeFromWire(w.value as never, exprFromWire as never),
    attr: nodeFromWire(w.attr as never, exprFromWire as never),
    ctx: w.ctx,
    hasQuestion: w.has_question === true,
  };
}

function callFromWire(w: Record<string, unknown>): unknown {
  return {
    func: nodeFromWire(w.func as never, exprFromWire as never),
    args: ((w.args as unknown[]) || []).map((a) => nodeFromWire(a as never, exprFromWire as never)),
    keywords: ((w.keywords as unknown[]) || []).map((k) => nodeFromWire(k as never, _dto.keywordFromWire as never)),
  };
}

function parenFromWire(w: Record<string, unknown>): unknown {
  return { expr: nodeFromWire(w.expr as never, exprFromWire as never) };
}

function quantFromWire(w: Record<string, unknown>): unknown {
  return {
    target: nodeFromWire(w.target as never, exprFromWire as never),
    variables: ((w.variables as unknown[]) || []).map((v) => nodeFromWire(v as never, exprFromWire as never)),
    op: w.op,
    test: nodeFromWire(w.test as never, exprFromWire as never),
    ifCond: nodeFromWire(w.if_cond as never, exprFromWire as never),
    ctx: w.ctx,
  };
}

function listFromWire(w: Record<string, unknown>): unknown {
  return {
    elts: ((w.elts as unknown[]) || []).map((e) => nodeFromWire(e as never, exprFromWire as never)),
    ctx: w.ctx,
  };
}

function listIfItemFromWire(w: Record<string, unknown>): unknown {
  return {
    ifCond: nodeFromWire(w.if_cond as never, exprFromWire as never),
    exprs: ((w.exprs as unknown[]) || []).map((e) => nodeFromWire(e as never, exprFromWire as never)),
    orelse: nodeFromWire(w.orelse as never, exprFromWire as never),
  };
}

function listCompFromWire(w: Record<string, unknown>): unknown {
  return {
    elt: nodeFromWire(w.elt as never, exprFromWire as never),
    generators: ((w.generators as unknown[]) || []).map((g) => nodeFromWire(g as never, exprFromWire as never)),
  };
}

function starredFromWire(w: Record<string, unknown>): unknown {
  return {
    value: nodeFromWire(w.value as never, exprFromWire as never),
    ctx: w.ctx,
  };
}

function dictCompFromWire(w: Record<string, unknown>): unknown {
  return {
    entryKey: nodeFromWire(w.entry_key as never, exprFromWire as never),
    key: nodeFromWire(w.key as never, exprFromWire as never),
    value: nodeFromWire(w.value as never, exprFromWire as never),
    generators: ((w.generators as unknown[]) || []).map((g) => nodeFromWire(g as never, exprFromWire as never)),
  };
}

function configIfEntryFromWire(w: Record<string, unknown>): unknown {
  return {
    ifCond: nodeFromWire(w.if_cond as never, exprFromWire as never),
    items: ((w.items as unknown[]) || []).map((i) => nodeFromWire(i as never, _dto.configEntryFromWire as never)),
    orelse: nodeFromWire(w.orelse as never, exprFromWire as never),
  };
}

function compClauseFromWire(w: Record<string, unknown>): unknown {
  return {
    targets: ((w.targets as unknown[]) || []).map((t) => nodeFromWire(t as never, exprFromWire as never)),
    iter: nodeFromWire(w.iter as never, exprFromWire as never),
    ifs: ((w.ifs as unknown[]) || []).map((i) => nodeFromWire(i as never, exprFromWire as never)),
  };
}

function schemaFromWire(w: Record<string, unknown>): unknown {
  return {
    name: nodeFromWire(w.name as never, exprFromWire as never),
    args: ((w.args as unknown[]) || []).map((a) => nodeFromWire(a as never, exprFromWire as never)),
    kwargs: ((w.kwargs as unknown[]) || []).map((k) => nodeFromWire(k as never, _dto.keywordFromWire as never)),
    config: nodeFromWire(w.config as never, exprFromWire as never),
  };
}

function configFromWire(w: Record<string, unknown>): unknown {
  return {
    items: ((w.items as unknown[]) || []).map((i) => nodeFromWire(i as never, _dto.configEntryFromWire as never)),
  };
}

function lambdaFromWire(w: Record<string, unknown>): unknown {
  return {
    args: nodeFromWire(w.args as never, _dto.argumentsFromWire as never),
    body: ((w.body as unknown[]) || []).map((b) => nodeFromWire(b as never, exprFromWire as never)),
    returnTy: nodeFromWire(w.return_ty as never, _types.typeFromWire as never),
  };
}

function subscriptFromWire(w: Record<string, unknown>): unknown {
  return {
    value: nodeFromWire(w.value as never, exprFromWire as never),
    index: nodeFromWire(w.index as never, exprFromWire as never),
    ctx: w.ctx,
  };
}

function compareFromWire(w: Record<string, unknown>): unknown {
  return {
    left: nodeFromWire(w.left as never, exprFromWire as never),
    ops: w.ops || [],
    comparators: ((w.comparators as unknown[]) || []).map((c) => nodeFromWire(c as never, exprFromWire as never)),
  };
}

function numberLitFromWire(w: Record<string, unknown>): unknown {
  return {
    binarySuffix: w.binary_suffix,
    value: w.value,
  };
}

function stringLitFromWire(w: Record<string, unknown>): unknown {
  return {
    isLongString: w.is_long_string === true,
    rawValue: w.raw_value || '""',
    value: w.value || "",
  };
}

function nameConstantLitFromWire(w: Record<string, unknown>): unknown {
  return { value: w.value };
}

function joinedStringFromWire(w: Record<string, unknown>): unknown {
  return {
    values: ((w.values as unknown[]) || []).map((v) => nodeFromWire(v as never, exprFromWire as never)),
  };
}

function formattedValueFromWire(w: Record<string, unknown>): unknown {
  return {
    value: nodeFromWire(w.value as never, exprFromWire as never),
    conversion: w.conversion,
    formatSpec: nodeFromWire(w.format_spec as never, exprFromWire as never),
  };
}

function missingFromWire(_w: Record<string, unknown>): unknown {
  return {};
}

const REGISTRY: Record<string, (w: Record<string, unknown>) => unknown> = {
  Target: targetFromWire,
  Identifier: identifierFromWire,
  Unary: unaryFromWire,
  Binary: binaryFromWire,
  If: ifFromWire,
  Selector: selectorFromWire,
  Call: callFromWire,
  Paren: parenFromWire,
  Quant: quantFromWire,
  List: listFromWire,
  ListIfItem: listIfItemFromWire,
  ListComp: listCompFromWire,
  Starred: starredFromWire,
  DictComp: dictCompFromWire,
  ConfigIfEntry: configIfEntryFromWire,
  CompClause: compClauseFromWire,
  Schema: schemaFromWire,
  Config: configFromWire,
  Lambda: lambdaFromWire,
  Subscript: subscriptFromWire,
  Compare: compareFromWire,
  NumberLit: numberLitFromWire,
  StringLit: stringLitFromWire,
  NameConstantLit: nameConstantLitFromWire,
  JoinedString: joinedStringFromWire,
  FormattedValue: formattedValueFromWire,
  Missing: missingFromWire,
};

export function exprFromWire(w: Record<string, unknown> | undefined | null): unknown {
  if (!w) return undefined;
  if (!w.type) return undefined;
  const variant = w.type as string;
  const loader = REGISTRY[variant];
  if (loader) return Object.assign({ type: variant }, loader(w));
  return { type: variant };
}

File: dotnet/KclLib.AST/Expr.cs
// Expr.cs — Expression hierarchy. Mirrors `ast::Expr` in `crates/ast/src/ast.rs`.
//
// Rust uses `#[serde(tag = "type")]` so each variant appears in JSON as
// `{"type": "<Variant>", ...}`. `Loaders.ExprFromWire` dispatches on that tag.

using System.Text.Json;

namespace KclLib.AST;

// We use records with `Type` as the discriminator. The `Type` field is
// populated by the dispatcher so downstream code can detect the variant
// even though C# records aren't class-tagged.

public record TargetExpr(string Type, NodeRef<string>? Name = null, List<MemberOrIndex>? Paths = null, string Pkgpath = "");
public record IdentifierExpr(string Type, List<NodeRef<string>>? Names = null, string Pkgpath = "", string? Ctx = null);
public record UnaryExpr(string Type, string? Op = null, NodeRef<object>? Operand = null);
public record BinaryExpr(string Type, NodeRef<object>? Left = null, string? Op = null, NodeRef<object>? Right = null);
public record IfExpr(string Type, NodeRef<object>? Body = null, NodeRef<object>? Cond = null, NodeRef<object>? Orelse = null);
public record SelectorExpr(string Type, NodeRef<object>? Value = null, NodeRef<object>? Attr = null, string? Ctx = null, bool HasQuestion = false);
public record CallExpr(string Type, NodeRef<object>? Func = null, List<NodeRef<object>>? Args = null, List<NodeRef<object>>? Keywords = null);
public record ParenExpr(string Type, NodeRef<object>? Expr = null);
public record QuantExpr(string Type, NodeRef<object>? Target = null, List<NodeRef<object>>? Variables = null, string? Op = null, NodeRef<object>? Test = null, NodeRef<object>? IfCond = null, string? Ctx = null);
public record ListExpr(string Type, List<NodeRef<object>>? Elts = null, string? Ctx = null);
public record ListIfItemExpr(string Type, NodeRef<object>? IfCond = null, List<NodeRef<object>>? Exprs = null, NodeRef<object>? Orelse = null);
public record ListComp(string Type, NodeRef<object>? Elt = null, List<NodeRef<object>>? Generators = null);
public record StarredExpr(string Type, NodeRef<object>? Value = null, string? Ctx = null);
public record DictComp(string Type, NodeRef<object>? EntryKey = null, NodeRef<object>? Key = null, NodeRef<object>? Value = null, List<NodeRef<object>>? Generators = null);
public record ConfigIfEntryExpr(string Type, NodeRef<object>? IfCond = null, List<NodeRef<object>>? Items = null, NodeRef<object>? Orelse = null);
public record CompClause(string Type, List<NodeRef<object>>? Targets = null, NodeRef<object>? Iter = null, List<NodeRef<object>>? Ifs = null);
public record SchemaExpr(string Type, NodeRef<object>? Name = null, List<NodeRef<object>>? Args = null, List<NodeRef<object>>? Kwargs = null, NodeRef<object>? Config = null);
public record ConfigExpr(string Type, List<NodeRef<object>>? Items = null);
public record LambdaExpr(string Type, NodeRef<Arguments>? Args = null, List<NodeRef<object>>? Body = null, NodeRef<object>? ReturnTy = null);
public record Subscript(string Type, NodeRef<object>? Value = null, NodeRef<object>? Index = null, string? Ctx = null);
public record Compare(string Type, NodeRef<object>? Left = null, List<string>? Ops = null, List<NodeRef<object>>? Comparators = null);
public record NumberLit(string Type, string? BinarySuffix = null, object? Value = null);
public record StringLit(string Type, bool IsLongString = false, string RawValue = "\"\"", string Value = "");
public record NameConstantLit(string Type, string? Value = null);
public record JoinedString(string Type, List<NodeRef<object>>? Values = null);
public record FormattedValue(string Type, NodeRef<object>? Value = null, string? Conversion = null, NodeRef<object>? FormatSpec = null);
public record MissingExpr(string Type);

internal static class ExprLoader
{
    public static object? ExprFromWire(JsonElement el)
    {
        if (el.ValueKind != JsonValueKind.Object) return null;
        var variant = WireHelpers.OptString(el, "type");
        if (variant == null) return null;
        return variant switch
        {
            "Target" => TargetFromWire(el),
            "Identifier" => IdentifierFromWire(el),
            "Unary" => UnaryFromWire(el),
            "Binary" => BinaryFromWire(el),
            "If" => IfFromWire(el),
            "Selector" => SelectorFromWire(el),
            "Call" => CallFromWire(el),
            "Paren" => ParenFromWire(el),
            "Quant" => QuantFromWire(el),
            "List" => ListFromWire(el),
            "ListIfItem" => ListIfItemFromWire(el),
            "ListComp" => ListCompFromWire(el),
            "Starred" => StarredFromWire(el),
            "DictComp" => DictCompFromWire(el),
            "ConfigIfEntry" => ConfigIfEntryFromWire(el),
            "CompClause" => CompClauseFromWire(el),
            "Schema" => SchemaFromWire(el),
            "Config" => ConfigFromWire(el),
            "Lambda" => LambdaFromWire(el),
            "Subscript" => SubscriptFromWire(el),
            "Compare" => CompareFromWire(el),
            "NumberLit" => NumberLitFromWire(el),
            "StringLit" => StringLitFromWire(el),
            "NameConstantLit" => NameConstantLitFromWire(el),
            "JoinedString" => JoinedStringFromWire(el),
            "FormattedValue" => FormattedValueFromWire(el),
            "Missing" => new MissingExpr("Missing"),
            _ => new BasicType(variant),
        };
    }

    private static TargetExpr TargetFromWire(JsonElement o)
    {
        List<MemberOrIndex>? paths = null;
        if (o.TryGetProperty("paths", out var pathsEl) && pathsEl.ValueKind == JsonValueKind.Array)
        {
            paths = new List<MemberOrIndex>();
            foreach (var p in pathsEl.EnumerateArray())
            {
                var mi = DtoLoader.MemberOrIndexFromWire(p);
                if (mi != null) paths.Add(mi);
            }
        }
        return new TargetExpr(
            Type: "Target",
            Name: WireHelpers.NodeFromWire<string>(o.GetProperty("name"), x => x.GetString()!),
            Paths: paths,
            Pkgpath: WireHelpers.OptString(o, "pkgpath") ?? ""
        );
    }

    private static IdentifierExpr IdentifierFromWire(JsonElement o) => new(
        Type: "Identifier",
        Names: WireHelpers.NodeListFromWire<string>(o.GetProperty("names"), x => x.GetString()!),
        Pkgpath: WireHelpers.OptString(o, "pkgpath") ?? "",
        Ctx: WireHelpers.OptString(o, "ctx")
    );

    private static UnaryExpr UnaryFromWire(JsonElement o) => new(
        Type: "Unary",
        Op: WireHelpers.OptString(o, "op"),
        Operand: WireHelpers.NodeFromWire<object>(o.GetProperty("operand"), Loaders.ExprFromWire)
    );

    private static BinaryExpr BinaryFromWire(JsonElement o) => new(
        Type: "Binary",
        Left: WireHelpers.NodeFromWire<object>(o.GetProperty("left"), Loaders.ExprFromWire),
        Op: WireHelpers.OptString(o, "op"),
        Right: WireHelpers.NodeFromWire<object>(o.GetProperty("right"), Loaders.ExprFromWire)
    );

    private static IfExpr IfFromWire(JsonElement o) => new(
        Type: "If",
        Body: WireHelpers.NodeFromWire<object>(o.GetProperty("body"), Loaders.ExprFromWire),
        Cond: WireHelpers.NodeFromWire<object>(o.GetProperty("cond"), Loaders.ExprFromWire),
        Orelse: WireHelpers.NodeFromWire<object>(o.GetProperty("orelse"), Loaders.ExprFromWire)
    );

    private static SelectorExpr SelectorFromWire(JsonElement o) => new(
        Type: "Selector",
        Value: WireHelpers.NodeFromWire<object>(o.GetProperty("value"), Loaders.ExprFromWire),
        Attr: WireHelpers.NodeFromWire<object>(o.GetProperty("attr"), Loaders.ExprFromWire),
        Ctx: WireHelpers.OptString(o, "ctx"),
        HasQuestion: WireHelpers.OptBool(o, "has_question")
    );

    private static CallExpr CallFromWire(JsonElement o) => new(
        Type: "Call",
        Func: WireHelpers.NodeFromWire<object>(o.GetProperty("func"), Loaders.ExprFromWire),
        Args: WireHelpers.NodeListFromWire<object>(o.GetProperty("args"), Loaders.ExprFromWire),
        Keywords: WireHelpers.NodeListFromWire<object>(o.GetProperty("keywords"), DtoLoader.KeywordFromWire)
    );

    private static ParenExpr ParenFromWire(JsonElement o) => new(
        Type: "Paren",
        Expr: WireHelpers.NodeFromWire<object>(o.GetProperty("expr"), Loaders.ExprFromWire)
    );

    private static QuantExpr QuantFromWire(JsonElement o) => new(
        Type: "Quant",
        Target: WireHelpers.NodeFromWire<object>(o.GetProperty("target"), Loaders.ExprFromWire),
        Variables: WireHelpers.NodeListFromWire<object>(o.GetProperty("variables"), Loaders.ExprFromWire),
        Op: WireHelpers.OptString(o, "op"),
        Test: WireHelpers.NodeFromWire<object>(o.GetProperty("test"), Loaders.ExprFromWire),
        IfCond: WireHelpers.NodeFromWire<object>(o.GetProperty("if_cond"), Loaders.ExprFromWire),
        Ctx: WireHelpers.OptString(o, "ctx")
    );

    private static ListExpr ListFromWire(JsonElement o) => new(
        Type: "List",
        Elts: WireHelpers.NodeListFromWire<object>(o.GetProperty("elts"), Loaders.ExprFromWire),
        Ctx: WireHelpers.OptString(o, "ctx")
    );

    private static ListIfItemExpr ListIfItemFromWire(JsonElement o) => new(
        Type: "ListIfItem",
        IfCond: WireHelpers.NodeFromWire<object>(o.GetProperty("if_cond"), Loaders.ExprFromWire),
        Exprs: WireHelpers.NodeListFromWire<object>(o.GetProperty("exprs"), Loaders.ExprFromWire),
        Orelse: WireHelpers.NodeFromWire<object>(o.GetProperty("orelse"), Loaders.ExprFromWire)
    );

    private static ListComp ListCompFromWire(JsonElement o) => new(
        Type: "ListComp",
        Elt: WireHelpers.NodeFromWire<object>(o.GetProperty("elt"), Loaders.ExprFromWire),
        Generators: WireHelpers.NodeListFromWire<object>(o.GetProperty("generators"), Loaders.ExprFromWire)
    );

    private static StarredExpr StarredFromWire(JsonElement o) => new(
        Type: "Starred",
        Value: WireHelpers.NodeFromWire<object>(o.GetProperty("value"), Loaders.ExprFromWire),
        Ctx: WireHelpers.OptString(o, "ctx")
    );

    private static DictComp DictCompFromWire(JsonElement o) => new(
        Type: "DictComp",
        EntryKey: WireHelpers.NodeFromWire<object>(o.GetProperty("entry_key"), Loaders.ExprFromWire),
        Key: WireHelpers.NodeFromWire<object>(o.GetProperty("key"), Loaders.ExprFromWire),
        Value: WireHelpers.NodeFromWire<object>(o.GetProperty("value"), Loaders.ExprFromWire),
        Generators: WireHelpers.NodeListFromWire<object>(o.GetProperty("generators"), Loaders.ExprFromWire)
    );

    private static ConfigIfEntryExpr ConfigIfEntryFromWire(JsonElement o) => new(
        Type: "ConfigIfEntry",
        IfCond: WireHelpers.NodeFromWire<object>(o.GetProperty("if_cond"), Loaders.ExprFromWire),
        Items: WireHelpers.NodeListFromWire<object>(o.GetProperty("items"), DtoLoader.ConfigEntryFromWire),
        Orelse: WireHelpers.NodeFromWire<object>(o.GetProperty("orelse"), Loaders.ExprFromWire)
    );

    private static CompClause CompClauseFromWire(JsonElement o) => new(
        Type: "CompClause",
        Targets: WireHelpers.NodeListFromWire<object>(o.GetProperty("targets"), Loaders.ExprFromWire),
        Iter: WireHelpers.NodeFromWire<object>(o.GetProperty("iter"), Loaders.ExprFromWire),
        Ifs: WireHelpers.NodeListFromWire<object>(o.GetProperty("ifs"), Loaders.ExprFromWire)
    );

    private static SchemaExpr SchemaFromWire(JsonElement o) => new(
        Type: "Schema",
        Name: WireHelpers.NodeFromWire<object>(o.GetProperty("name"), Loaders.ExprFromWire),
        Args: WireHelpers.NodeListFromWire<object>(o.GetProperty("args"), Loaders.ExprFromWire),
        Kwargs: WireHelpers.NodeListFromWire<object>(o.GetProperty("kwargs"), DtoLoader.KeywordFromWire),
        Config: WireHelpers.NodeFromWire<object>(o.GetProperty("config"), Loaders.ExprFromWire)
    );

    private static ConfigExpr ConfigFromWire(JsonElement o) => new(
        Type: "Config",
        Items: WireHelpers.NodeListFromWire<object>(o.GetProperty("items"), DtoLoader.ConfigEntryFromWire)
    );

    private static LambdaExpr LambdaFromWire(JsonElement o) => new(
        Type: "Lambda",
        Args: WireHelpers.NodeFromWire<Arguments>(o.GetProperty("args"), DtoLoader.ArgumentsFromWire!),
        Body: WireHelpers.NodeListFromWire<object>(o.GetProperty("body"), Loaders.ExprFromWire),
        ReturnTy: WireHelpers.NodeFromWire<object>(o.GetProperty("return_ty"), Loaders.TypeFromWire)
    );

    private static Subscript SubscriptFromWire(JsonElement o) => new(
        Type: "Subscript",
        Value: WireHelpers.NodeFromWire<object>(o.GetProperty("value"), Loaders.ExprFromWire),
        Index: WireHelpers.NodeFromWire<object>(o.GetProperty("index"), Loaders.ExprFromWire),
        Ctx: WireHelpers.OptString(o, "ctx")
    );

    private static Compare CompareFromWire(JsonElement o) => new(
        Type: "Compare",
        Left: WireHelpers.NodeFromWire<object>(o.GetProperty("left"), Loaders.ExprFromWire),
        Ops: WireHelpers.OptList<string>(o, "ops"),
        Comparators: WireHelpers.NodeListFromWire<object>(o.GetProperty("comparators"), Loaders.ExprFromWire)
    );

    private static NumberLit NumberLitFromWire(JsonElement o) => new(
        Type: "NumberLit",
        BinarySuffix: WireHelpers.OptString(o, "binary_suffix"),
        Value: o.TryGetProperty("value", out var v) ? JsonSerializer.Deserialize<object>(v.GetRawText()) : null
    );

    private static StringLit StringLitFromWire(JsonElement o) => new(
        Type: "StringLit",
        IsLongString: WireHelpers.OptBool(o, "is_long_string"),
        RawValue: WireHelpers.OptString(o, "raw_value") ?? "\"\"",
        Value: WireHelpers.OptString(o, "value") ?? ""
    );

    private static NameConstantLit NameConstantLitFromWire(JsonElement o) => new(
        Type: "NameConstantLit",
        Value: WireHelpers.OptString(o, "value")
    );

    private static JoinedString JoinedStringFromWire(JsonElement o) => new(
        Type: "JoinedString",
        Values: WireHelpers.NodeListFromWire<object>(o.GetProperty("values"), Loaders.ExprFromWire)
    );

    private static FormattedValue FormattedValueFromWire(JsonElement o) => new(
        Type: "FormattedValue",
        Value: WireHelpers.NodeFromWire<object>(o.GetProperty("value"), Loaders.ExprFromWire),
        Conversion: WireHelpers.OptString(o, "conversion"),
        FormatSpec: WireHelpers.NodeFromWire<object>(o.GetProperty("format_spec"), Loaders.ExprFromWire)
    );
}